ml-agents/mlagents/trainers/trainer_util.py

```diff
@@ -38,6 +40,7 @@ def __init__(
         self.seed = seed
         self.meta_curriculum = meta_curriculum
         self.multi_gpu = multi_gpu
+        self.ghost_controller = GhostController(ghost_swap)
```

Comment:
It seems GhostController 1:1 with a trainer. Can you explain how this makes it work for asymmetric use cases (multiple trainers)
Do we always need a GhostController even when there is a single agent ? It seems we are creating an object we might not always need.

Reply:
GhostController is not 1:1 with a trainer.  A reference to the ghost_controller is given to each ghost trainer when it's created.  On every advance, the ghost trainers query the GhostController for the learning team (only one team is learning at a time). The trainers pass its current training step as part of the query. When the training step exceeds the swap threshold, the next team in the queue becomes the learning team.  All the other trainers get this new team upon querying.

The trainer knows which team id is learning. The ghost trainer simply iterates through its queues. If the trajectory queue/policy queue match the learning team, it passes the trajectory to the wrapped trainer and updated policy to the environment.  Otherwise, it ignores trajectories and sends ghost policies to the environment.

Reply:
To your second point, yes it's always created and no it's not used unless there are ghost trainers.  I will figure out a better way to do this.